### PR TITLE
RUN: Run rustfmt on the document before saving

### PR DESCRIPTION
**Pull request by mchernyavsky**
`Preferences > Languages & Frameworks > Rust > Rustfmt > Run rustfmt on Save`

Also, now rustfmt uses the `rustfmt.toml` from the corresponding cargo project root.

Relates to https://github.com/intellij-rust/intellij-rust/issues/1087.

### CODE REVIEW

**Comment by vlad20012** [comment]
I don't think #1087 should be closed by this PR. It's just a part

**Comment by brokenthorn** [comment]
Awesome! Thank you @mchernyavsky!

**Comment by aldanor** [comment]
This is pretty awesome, thanks.

One question re: rustfmt.toml: will this work in workspace projects? (where rustfmt.toml is typically located at the workspace root, not child crate root)

**Comment by aldanor** [comment]
I know this is probably outside of the scope of this PR, but since it touches rustfmt and its config, I'll mention it anyway. Would it be viable to separate rustfmt channel from project channel? (E.g. compile the project and run its tests on stable but use +nightly for rustfmt) There's some rustfmt options that are still waiting to be stabilized and you have to use nightly to make use of them.

**Comment by mchernyavsky** [comment]
>One question re: rustfmt.toml: will this work in workspace projects? (where rustfmt.toml is typically located at the workspace root, not child crate root)

Good point! I have fix it in separate PR (https://github.com/intellij-rust/intellij-rust/pull/3506).

>Would it be viable to separate rustfmt channel from project channel?

I'm working on https://github.com/intellij-rust/intellij-rust/issues/3318 now. This "default project toolchain" will also be used by rustfmt.

File: src/main/kotlin/org/rust/openapiext/CommandLineExt.kt [hunk 1]
```diff
@@ -80,6 +81,13 @@ fun GeneralCommandLine.execute(
     }
 
     listener?.let { handler.addProcessListener(it) }
+
+    if (stdIn != null) {
+        val processInput = handler.processInput
+        processInput?.write(stdIn)
+        processInput?.close()
```
Comment (Undin) [inline src/main/kotlin/org/rust/openapiext/CommandLineExt.kt]: Let's replace with `handler.processInput?.use { it.write(stdIn) }`

File: src/main/kotlin/org/rust/ide/RustfmtWatcher.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import com.intellij.openapi.project.guessProjectForFile
+import com.intellij.util.containers.ContainerUtil
+import org.rust.cargo.project.settings.rustSettings
+import org.rust.cargo.project.settings.toolchain
+import org.rust.openapiext.virtualFile
+
+class RustfmtWatcher : BaseComponent {
+    private val documentsToReformatLater: MutableSet<Document> = ContainerUtil.newConcurrentSet()
+    @Volatile
+    var isEnabled: Boolean = true
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/RustfmtWatcher.kt]: Why do we need volatile here?
We always use (and looks like must use) it in EDT

File: src/main/kotlin/org/rust/ide/RustfmtWatcher.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                rustfmt.reformatDocument(project, document)
+            }
+        }
+
+        override fun beforeDocumentSaving(document: Document) {
+            if (!isEnabled) return
+            val project = guessProjectForFile(document.virtualFile) ?: return
+            if (!project.rustSettings.runRustfmtOnSave) return
+            val rustfmt = project.toolchain?.rustfmt() ?: return
+            rustfmt.reformatDocument(project, document)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/RustfmtWatcher.kt]: This code is the same as in `beforeAllDocumentsSaving`
Maybe extract it into a separate method?

File: src/main/kotlin/org/rust/ide/RustfmtWatcher.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        if (!project.rustSettings.runRustfmtOnSave) return false
+        return documentsToReformatLater.add(document)
+    }
+
+    private inner class RustfmtListener : FileDocumentManagerListener {
+        override fun beforeAllDocumentsSaving() {
+            val documentsToReformat = HashSet(documentsToReformatLater)
+            documentsToReformatLater.clear()
+            for (document in documentsToReformat) {
+                val project = guessProjectForFile(document.virtualFile) ?: continue
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/RustfmtWatcher.kt]: Maybe check rust file type here before project search?
Now we run `rustfmt` for each file even it isn't rust file => launch redundant external processes

File: src/main/kotlin/org/rust/ide/annotator/cargoCheck/RsCargoCheckUtils.kt [hunk 5]
```diff
@@ -320,15 +317,21 @@ private fun formatMessage(message: String): String {
  * these documents should be stripped later (when [saveAllDocuments] is called).
  */
 private fun saveAllDocumentsAsTheyAre() {
+    checkWriteAccessAllowed()
     val documentManager = FileDocumentManager.getInstance()
-    for (document in documentManager.unsavedDocuments) {
+    val rustfmtWatcher = ApplicationManager.getApplication().getComponent(RustfmtWatcher::class.java)
+    val rustfmtWatcherStatus = rustfmtWatcher.isEnabled
+    rustfmtWatcher.isEnabled = false
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/cargoCheck/RsCargoCheckUtils.kt]: Maybe it's better to create `RustfmtWatcher#withoutReformating` method which takes lambda and move all logic related to changes in `RustfmtWatcher` state into `RustfmtWatcher` itself
It should provide a simpler and prettier api

File: src/main/kotlin/org/rust/ide/annotator/cargoCheck/RsCargoCheckUtils.kt [hunk 6]
```diff
@@ -320,15 +317,21 @@ private fun formatMessage(message: String): String {
...
+    val (pre, post) = documentManager.unsavedDocuments.partition { it.virtualFile?.name in Rustfmt.CONFIG_FILES }
+    for (document in pre + post) {
         documentManager.saveDocumentAsIs(document)
         documentManager.stripDocumentLater(document)
+        rustfmtWatcher.reformatDocumentLater(document)
     }
+    rustfmtWatcher.isEnabled = rustfmtWatcherStatus
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/cargoCheck/RsCargoCheckUtils.kt]: I'd prefer to wrap all code between changing `rustfmtWatcher.isEnabled` value to `try/finally` block to be sure that the initial value is restored even in case of exception

**Comment by mchernyavsky** [comment]
@Undin fixed.

File: src/main/kotlin/org/rust/cargo/runconfig/command/RunClippyAction.kt [hunk 7]
```diff
@@ -5,47 +5,17 @@
...
-        )
-
-        if (needInstallClippy) {
-            val action = InstallComponentAction(projectDirectory, "clippy-preview")
-            project.showBalloon("Clippy is not installed", NotificationType.ERROR, action)
-            return
-        }
-
-        CargoCommandLine.forProject(cargoProject, "clippy").run(project, cargoProject)
+        checkNeedInstallClippy(cargoProject)
```
Comment (Undin) [inline src/main/kotlin/org/rust/cargo/runconfig/command/RunClippyAction.kt]: The previous version runs `clippy` existence under progress, the current doesn't
Could we revert this change?

File: src/main/kotlin/org/rust/cargo/toolchain/Rustup.kt [hunk 8]
```diff
@@ -81,4 +87,43 @@ class Rustup(
         val fs = LocalFileSystem.getInstance()
         return fs.refreshAndFindFileByPath(FileUtil.join(sysroot, "lib/rustlib/src/rust/src"))
     }
+
+    companion object {
+
+        // We don't want to install Clippy if:
+        // 1. It is already installed
+        // 2. We don't have Rustup
+        // 3. Rustup doesn't have Clippy component
```
Comment (Undin) [inline src/main/kotlin/org/rust/cargo/toolchain/Rustup.kt]: Let's move this comment into `checkNeedInstallComponent` and generalize its text

File: src/main/kotlin/org/rust/ide/actions/RustfmtFileAction.kt [hunk 9]
```diff
@@ -33,29 +31,22 @@ class RustfmtFileAction : DumbAwareAction() {
     override fun actionPerformed(e: AnActionEvent) {
         val (project, toolchain, file) = getContext(e) ?: return
 
-        FileDocumentManager.getInstance().saveAllDocuments()
+        val cargoProject = project.cargoProjects.findProjectForFile(file) ?: return
+        if (checkNeedInstallRustfmt(cargoProject)) return
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/actions/RustfmtFileAction.kt]: The same comment about lost progress

**Comment by mchernyavsky** [comment]
Fixed.

**Comment by Undin** [comment]
@mchernyavsky Unfortunately, now it conflicts with changes from your other PR

**Comment by mchernyavsky** [comment]
@Undin Done.
